schema/pgsql/schema.sql

```diff
@@ -278,3 +278,32 @@ CREATE TABLE incident_history (
     CONSTRAINT pk_incident_history PRIMARY KEY (id),
     FOREIGN KEY (incident_id, rule_escalation_id) REFERENCES incident_rule_escalation_state(incident_id, rule_escalation_id)
 );
+
+CREATE OR REPLACE FUNCTION now_extended() RETURNS bigint
+    LANGUAGE SQL AS
+$$
+SELECT ((EXTRACT(EPOCH FROM CURRENT_TIMESTAMP) * 1_000) :: bigint);
+$$;
+
+CREATE TABLE session
+(
+    id               VARCHAR(256) NOT NULL,
+    username         VARCHAR(254) NOT NULL,
+    device_id        VARCHAR(8)   NOT NULL,
```

Comment:
`id` stores the session identificator generated by php. We use VARCHAR(256) here as that's the maximum number of allowed characters stated by PHP's [documentation](https://www.php.net/manual/en/session.configuration.php#ini.session.sid-length).

`username`'s type was defined by double-checking the existing username table definitions found in [icingaweb2/schema/pgsql.schema.sql](https://github.com/Icinga/icingaweb2/blob/main/schema/pgsql.schema.sql#L38C27-L38C27).

`device_id` stores a device fingerprint generated by creating a [jooat hash](https://en.wikipedia.org/wiki/Jenkins_hash_function#one_at_a_time) which consinsts of the authenticated username, concatenated by the full user-agent string. Jooat hashes have a maximum length of 8 bytes, hence the limitation.

Reply:
This comment is now pretty much a "I think it's strange but if you want it, you get it" as per https://github.com/Icinga/icinga-notifications/issues/118#issuecomment-1780964049.

For `username`, I find it a bit weird that it uses a different type here compared to `contact.username`. Though that one might be subject to a change to `citext` in #71 based on feedback from @nilmerg on how Icinga Web handles usernames.

> `device_id` stores a device fingerprint generated by creating a [jooat hash](https://en.wikipedia.org/wiki/Jenkins_hash_function#one_at_a_time) which consinsts of the authenticated username, concatenated by the full user-agent string. Jooat hashes have a maximum length of 8 bytes, hence the limitation.

`uint32_t` vs `varchar(8)`, sounds like you want to store an int value as a hex string? And that hash doesn't really sound like it would give something that would qualify as device ID: if I use the same browser on the same OS on two devices, this should give the same device ID whereas if I update my browser on one device, that would change the device ID.

Reply:
`username` uses the type definition from [icingaweb2/schema/pgsql.schema](https://github.com/Icinga/icingaweb2/blob/main/schema/pgsql.schema.sql) as this is the actual type representation of the username  that's being passed in the AuthenticationHook parameters. As we're storing the session and its corresponding user in that specific hook, we also want to use the same type definition (I'd say).
You are right with the `uint32_t` that's used as a return value in jenkin's one at a time hash. We are working with PHP however and a call to [hash()](https://www.php.net/manual/en/function.hash.php) returns lowercased hexits by default. It most certainly makes sense to store the uint representation of that value but shouldn't the database reflect the data representation that's also getting used in comparisons and such?
Because in the written PHP code, it will always be compared in its hexadecimal form.

You have a fair point in terms of the `device_id` column naming convention. It essentially fingerprints a specific browser and its authenticated user. It doesn't really care about the underlying device.
I chose that prefix as other features in the icinga2 suite which also use the user agent string, already refer to it as a device, e.g: [icingaweb2/library/Icinga/Web/RememberMeUserDevicesList.php](https://github.com/Icinga/icingaweb2/blob/main/library/Icinga/Web/RememberMeUserDevicesList.php#L77)

Reply:
@julianbrost updated the code to include your requested changes.
Is this fine for you now?